src/app/icd/client/CheckInHandler.cpp

```diff
@@ -0,0 +1,191 @@
+    VerifyOrReturnError(clientStorage != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mpExchangeManager == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mpICDClientStorage == nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeManager  = exchangeManager;
+    mpICDClientStorage = clientStorage;
+    mpCheckInDelegate  = delegate;
+
+    return mpExchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn, this);
+}
+
+void CheckInHandler::Shutdown()
+{
+    mpICDClientStorage = nullptr;
+    mpCheckInDelegate  = nullptr;
+    if (mpExchangeManager)
+    {
+        mpExchangeManager->UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn);
+        mpExchangeManager = nullptr;
+    }
+}
+
+CHIP_ERROR CheckInHandler::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate)
+{
+    // Return error for wrong message type
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn),
+                        CHIP_ERROR_INVALID_MESSAGE_TYPE);
+
+    newDelegate = this;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckInHandler::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                             System::PacketBufferHandle && payload)
+{
+    // If the message type is not ICD_CheckIn, return CHIP_NO_ERROR and exit
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn), CHIP_NO_ERROR);
+
+    ByteSpan payloadByteSpan{ payload->Start(), payload->DataLength() };
+    ICDClientInfo clientInfo;
+    CounterType counter = 0;
+    // If the check-in message processing fails, return CHIP_NO_ERROR and exit.
+    VerifyOrReturnError(CHIP_NO_ERROR == mpICDClientStorage->ProcessCheckInPayload(payloadByteSpan, clientInfo, counter),
+                        CHIP_NO_ERROR);
+    CounterType receivedCheckInCounterOffset = (counter - clientInfo.start_icd_counter) % kCheckInCounterMax;
+
+    // Detect duplicate check-in messages and return CHIP_NO_ERROR on receiving a duplicate message
+    if (receivedCheckInCounterOffset <= clientInfo.offset)
+    {
+        ChipLogError(ICD, "A duplicate check-in message was received and discarded");
+        return CHIP_NO_ERROR;
+    }
+
+    clientInfo.offset = receivedCheckInCounterOffset;
+    bool refreshKey   = (receivedCheckInCounterOffset > kKeyRefreshLimit);
+
+    if (refreshKey)
+    {
+        uint8_t newKeyData[chip::Crypto::kAES_CCM128_Key_Length];
+        mpCheckInDelegate->OnRefreshKeyGenerate(clientInfo, newKeyData, chip::Crypto::kAES_CCM128_Key_Length);
+        // A new session should be established to re-register the client using the new key. The registration will happen in
+        // mOnDeviceConnected callback
+        EstablishSessionToPeer(clientInfo.peer_node);
```

Comment:
So what I would think we should do here is:
1. Establish the session.
2. When the session is established, ask the delegate for a new key and use it to re-register.

That avoids the whole "generate the key now, but have it get stored by the delegate until we actually need it later".  Why not just generate it when we need it?

Reply:
Or is the idea that after we establish the session we don't have the ICDClientInfo anymore?  But do we really need it?  Just the ScopedNodeId should be fine for purposes of "please give me a key, and if you want to remember what key is used for which node, it's for this one", no?

Reply:
The idea was to initiate registration from the delegate. For registration, we need monitored subject from ICDClientInfo along with the new key and ScopedNodeId. That was the reason to store and retrieve the key with ICDClientInfo. Do you recommend a better way for this?

Reply:
Discussed offline. Will implement the new design in a follow-up PR.